feat: Update the data export process to include extra data files (M2-8485)

sultanofcardio opened the pull request:
- [x] Tests for the changes have been added
- [ ] OSS packages added to MindLogger [open source credit page](https://mindlogger.atlassian.net/wiki/spaces/MINDLOGGER1/pages/263127186/Admin+Panel+Applet+Builder+Library)

### 📝 Description

🔗 [Jira Ticket M2-8485](https://mindlogger.atlassian.net/browse/M2-8485)

This PR introduces two new file exports:
- Schedule history
- Flow item history

These new files should be present everywhere exporting data is currently available in the admin app. Including:
- Dashboard respondents tab (`/dashboard/respondents`) using the applet context menu
- Applet page top-level export button (e.g. on the overview page at `/dashboard/{applet_id}/overview`)
- Applet page activities tab (`/dashboard/{applet_id}/activities`) using the activity/flow context menu
- Applet page participants tab (`/dashboard/{applet_id}/participants`) using the participant table row context menu
- Participant detail page assignments tab (`/dashboard/{applet_id}/participants/{subject_id}`) using the activity/flow/participant context menu
- DataViz Summary tab (`/dashboard/{applet_id}/participants/{subject_id}/activities/{activity_id}/summary`) export button
- Applet Settings (`/dashboard/{applet_id}/settings`) export data button

> [!NOTE]
> The files will only be exported when the `enable-ema-extra-files` feature flag is set to `true`

#### Flow Item History

Given a date range, a file called `flow_activity_history.csv` is produced containing the history of each version of each activity within each version of each activity flow inside a given applet, even if they have since been deleted.

Columns:
- `applet_id`: Applet ID
- `flow_item_history_created_date`: Date that the flow item was added to the flow
- `activity_flow_id`: Flow ID
- `activity_flow_name`: Flow Name
- `applet_version`: Applet version number
- `activity_id`: Activity ID
- `activity_name`: Activity Name

> [!NOTE]
> See the [ticket](https://mindlogger.atlassian.net/browse/M2-8485) description for a sample `flow_activity_history.csv` file

#### Schedule History

Given a date range, a new file called `schedule_history.csv` is produced containing an enumeration of the history of scheduled events (both default and individual) between the start and end of the date range (inclusive) for every user in a given applet. An enumeration here means that for every day in the range (inclusive), we use the periodicity of each event version to project its availability on that given day for every user in the applet. This is a look back in time to approximate which days in the period each event version was available to the user.

For example, given the date range 2025-03-09 to 2025-03-15, we will look back in time at the following days to determine which events apply to these days. Days with no events will not show up in the file
- 2025-03-09
- 2025-03-10
- 2025-03-11
- 2025-03-12
- 2025-03-13
- 2025-03-14
- 2025-03-15

Using this example, if an ALWAYS available event was created on 2025-03-13,  there should be rows in the file for:
- 2025-03-13
- 2025-03-14
- 2025-03-15. 

However, a WEEKDAYS available event created on the same day (2025-03-13) will only have rows for
- 2025-03-13
- 2025-03-14

Columns:
- `applet_id`: Applet ID
- `applet_version`: Applet version number
- `user_id`: User ID
- `target_or_source_id`: The subject ID of the user identified by user_id
- `schedule_id`: Event ID
- `schedule_type`: activity or flow
- `schedule_version`: Event version number in the format YYYYMMDD-n, where n is the nth change on a given date
- `schedule_version_created_timestamp`: The time when this version of the event was created in UTC timezone
- `schedule_updated_by`: The user ID of the logged-in admin user who updated the event and created this version
- `activity_or_flow_id`: Activity ID or Flow ID based on what’s in event_type
- `activity_or_flow_name`: Activity name or Flow name based on what’s in event_type
- `activity_or_flow_hidden`: Whether the activity or flow is hidden in this version of the applet (values yes or no)
- `schedule_start_date`: Start date of the event periodicity
- `schedule_end_date`: End date of the event periodicity
- `schedule_date`: The given date within the period
- `schedule_start_time`: Event start time on a given date within the period
- `schedule_end_time`: Event end time on a given date within the period
- `access_before_schedule_start_time`: Whether the admin has enabled “Access before from time” on the event (values yes or no)
- `mobile_device_schedule_version`: The latest version of this event that this user has on any of their mobile devices
- `mobile_device_schedule_start_date`: Start date of the event periodicity from the version of the event that the user has on their mobile devices
- `mobile_device_schedule_end_date`: End date of the event periodicity from the version of the event that the user has on their mobile devices
- `mobile_device_schedule_start_time`: The start time from the version of the event that the user has on their mobile devices
- `mobile_device_schedule_end_time`: The end time from the version of the event that the user has on their mobile devices
- `mobile_device_schedule_access_before_start_time`: The value of the “Access before from time” flag on the version of the event on the user’s mobile device (values yes or no)
- `mobile_device_schedule_download_timestamp`: The date and time when the user downloaded this version of the event in UTC timezone
- `mobile_device_schedule_utc_timezone_offset`: The number of minutes to be added to the mobile_device_schedule_download_timestamp to convert it to the user’s timezone (for comparison with event start and end times)

> [!NOTE]
> See the [ticket](https://mindlogger.atlassian.net/browse/M2-8485) description for a sample `schedule_history.csv` file

This is an incredibly complicated endeavour with a lot of scenarios and edge cases. I have tried to flesh some of these out in a test file that hasn't been committed yet. However, the following scenarios have been identified as some important ones for the exported files:

- the schedule changes on the server **before** the previously scheduled event, and the device gets it immediately
- the schedule changes on the server **after** the previously scheduled event, and the device gets it immediately
- the schedule changes on the server **during** the previously scheduled event, and the device gets it immediately
- the schedule changes on the server **before** the previously scheduled event, and the device gets it **after the end, but still on the same day** of the previously scheduled event
- the schedule changes on the server **before** the previously scheduled event, and the device gets it **during** of the previously scheduled event
- the schedule changes on the server **before** the previously scheduled event, and the device gets it **after the end of the previously scheduled event...in fact, several days later**

> [!NOTE]
> See [MindLogger Schedule History Export](https://docs.google.com/spreadsheets/d/1lHWiwN1__e0j6REMW6ylY4wTxeG9We4UOlBINePr8W0/edit?gid=898155703#gid=898155703) Google Sheets file for samples of these scenarios

### 📸 Screenshots

N/A

### 🪤 Peer Testing

N/A

### ✏️ Notes

1. I have decided to use the `luxon` library for date/time calculations as part of this PR because I'm more familiar with it and it was already being included in the project as a transitive dependency. Usually I would have ultimately converted these calculations back to our current library `date-fns`, but I have decided not to do that in the interest of time.

sultanofcardio committed:
Implement ScheduleHistoryExporter and FlowItemHistoryExporter

sultanofcardio committed:
Fix schedule history logic

The logic that determines whether an older event has been superseded was checking for the start of the day instead of the start of the event

mbanting commented on src/modules/Dashboard/features/Respondents/Popups/DataExportPopup/DataExportPopup_old.tsx:
```diff
@@ -135,6 +139,26 @@ export const DataExportPopup = ({
+              toDate,
+            });
+          } else {
+            console.warn(
+              `No owner ID found for current workspace. Schedule history export skipped.`,
+            );
```
In what cases does this occur, out of curiosity?

sultanofcardio replied:
I guess if the redux workspace data fails to be populated

<img width="509" alt="image" src="https://github.com/user-attachments/assets/1edae1ca-ce82-40ae-a730-3181d97127bd" />

So technically it should never happen, since if it did then the applet wouldn't load to get to this point. This is just me handling the type error

mbanting replied:
Understood, thanks!

mbanting commented on src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts:
```diff
@@ -0,0 +1,399 @@
+  respondentIds?: string;
+};
+
+export class ScheduleHistoryExporter extends DataExporter<
```
Unlike `FlowItemHistoryExporter.ts` this is more complex and not as obvious what this is exporting. Recommend a brief description similar to what you provided in this PR description so that it's captured as part of the code, something like "...exports an enumeration of the history of scheduled events (both default and individual) between the start and end of the date range (inclusive) for every user in a given applet. "

sultanofcardio replied:
Good call. I still have a bit of cleanup to do. Thanks for the suggestion 🙏

mbanting commented on src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts:
```diff
@@ -0,0 +1,399 @@
+      params?.fromDate ?? sortedDeviceScheduleHistoryData[0].createdAt,
+    );
+    // const periodEndDate = new Date(params?.toDate ?? DateTime.now().toISO());
+    const periodEndDate = DateTime.fromISO('2025-03-14').toJSDate();
```
Leftover mock data?

sultanofcardio replied:
Indeed. Thanks for noticing this

mbanting commented on src/shared/utils/exportData/exporters/ScheduleHistoryExporter.ts:
```diff
@@ -0,0 +1,399 @@
+            sortedDeviceScheduleHistoryData,
+          );
+
+          // TODO: Figure out which schedule applies here, individual vs default
```
Is this still to be done?

sultanofcardio replied:
At the time, yes. I've since addressed this as one of the outstanding items while writing the test cases

mbanting reviewed (changes requested):
Took a first pass at the code. A challenging task nicely broken down and organized. Left a couple minor of comments, and I know the tests are still being written. Thank you as well for the extensive PR description. Super helpful. Great job!

sultanofcardio committed:
Consider the end date for weekly and monthly periodicity

sultanofcardio committed:
Remove temporary test data

sultanofcardio committed:
Remove default schedules when individual schedules should apply

sultanofcardio committed:
Add WIP ScheduleHistoryExporter test file

sultanofcardio committed:
WIP update to ScheduleHistoryExporter test file

sultanofcardio committed:
Fix logic that replaces old event version with new one

sultanofcardio committed:
Update the logic that includes a default schedule after individual schedule deletion

When an individual schedule is deleted, the default schedule starts applying again. Whether it applies on the deletion depends on the time of deletion of the individual schedule event:

- Deletion before the end time of the default schedule means it applies on that date (depending on periodicity).
- Deletion after the end time means it does not apply

Similarly, whether the individual schedule that was deleted applies on its deletion date depends on the time of its deletion:

- Deletion before its start time means the event does not apply for that day (depending on periodicity)
- Deletion after its start time means it applies

So it is possible to have both the individual schedule and default schedule events applying on the day of deletion of the individual schedule event

sultanofcardio committed:
Add a few more tests

sultanofcardio committed:
Fix comparison of individual schedule event deletion time to default schedule event end time

sultanofcardio committed:
Rename variable

sultanofcardio committed:
Clean up test cases

sultanofcardio committed:
Finish up tests for findSchedulesForDays

sultanofcardio committed:
Add documentation to each class

sultanofcardio committed:
Exclude default schedule events for separate activities when any individual schedule is active

sultanofcardio committed:
Add test case for shorter months in monthly periodicity

sultanofcardio committed:
Merge branch 'develop' into feat/M2-8485-ema-data-exports

sultanofcardio committed:
Include whether the activity/flow is hidden in the schedule history file

sultanofcardio committed:
Account for when an event's end time extends past the end of the day

sultanofcardio committed:
Group events by applet version before processing

sultanofcardio commented:
I've updated this PR to do a few additional things:
- Indicate whether an activity/flow is hidden in the export file (this one depends on https://github.com/ChildMindInstitute/mindlogger-backend-refactor/pull/1787)
- Account for events that extend past the end of the day (e.g. start at 10pm today and end at 2am tomorrow)
- Group events by applet version to avoid comparison between otherwise identical events

sultanofcardio committed:
Account for events that allow access before the start time

These events actually allow access from midnight

sultanofcardio committed:
Add 3 new columns to the schedule export file

They are:
- access_before_schedule_start_time
- mobile_device_schedule_access_before_start_time
- mobile_device_schedule_utc_timezone_offset

sultanofcardio committed:
Add schedule export records in descending order

sultanofcardio committed:
Include end date in the daysBetweenInterval function

sultanofcardio committed:
Fix default from date

sultanofcardio committed:
Remove redundant start and end date variables

sultanofcardio committed:
Only include mobile schedules downloaded before the start time on the day

sultanofcardio committed:
Don't enumerate events before their creation date

sultanofcardio committed:
Fix interval generation

Turns out we're setting the testing timezone to UTC so there are some cases that weren't showing up before

sultanofcardio committed:
Rename some variables to be clearer

sultanofcardio committed:
Restrict applet versions in the export file

Applet versions in the export file will now only appear for the periods after which they were created, and before being replaced by a new applet version.

The consequence is that we can no longer project data before schedule history was being collected in the database

sultanofcardio committed:
Add missing target subject IDs argument

mbanting reviewed (approved):
Focused my re-review mostly on `ScheduleHistoryExporter`, it's tests, and the updated `daysBetweenInterval` function.

Very comprehensive implementation driven and validated by the impressive number of scenarios and their tests. Impressive!

sultanofcardio committed:
Update tests

sultanofcardio committed:
Remove comment

sultanofcardio committed:
Account for user timezone when finding latest mobile schedule

sultanofcardio committed:
Fix logic that determines the exclusion of events in the past based on applet version

sultanofcardio committed:
Make schedule exporter more testable

I extracted the row generation logic into a method that returns the rows so that we can make assertions without intercepting the file download.

I also abstracted the data requests into their own methods

sultanofcardio committed:
Set autogenerated date range to UTC

The dates sent from the export modal are in UTC, so we need to standardize on UTC for the autogenerated dates and the dates in the export file

sultanofcardio committed:
Filter respondents by subject ID

This limits th export file to only the subjects whose data is being exported

sultanofcardio committed:
Fix mobile schedule comparison logic

When finding the latest downloaded mobile schedule, we are currently comparing the version numbers incorrectly. This commit fixes the logic to only compare the `-n` part of the version numbers if the date parts are the same

sultanofcardio committed:
Change daysBetweenInterval function args

This functions now accepts string args in the format yyyy-MM-dd instead of a luxon DateTime object. This should make it easier to consume and there should be less ambiguity about the two ends of the date range

sultanofcardio committed:
Update how individual schedules override default ones

Default schedules will now always show up on the day when an individual schedule was created. Similarly, it will always show up on the day an individual schedule was deleted

sultanofcardio committed:
Update how new schedule versions override previous ones

Whenever a new version of an event is created, or a new event is created that flips the entity between being scheduled and always available, the previous version will now always show up in the export file on the creation day

sultanofcardio committed:
Include deleted events on their deletion dates

sultanofcardio committed:
Update applet version override logic

An old applet version will now always show up on the day the next version is created, regardless of the time. Similarly, a new applet version will always appear on its creation day, regardless of the time it was created

sultanofcardio committed:
Remove outdated comment

sultanofcardio committed:
Update overriding individual schedule test cases

sultanofcardio committed:
Remove incomplete test case

sultanofcardio committed:
Update deleted individual schedule test cases

sultanofcardio committed:
Update new schedule version test cases

sultanofcardio committed:
Update new schedule instance test cases

These are the test cases where scheduled access is replaced by always available and vice versa

sultanofcardio committed:
Update missing individual schedule event test case

sultanofcardio committed:
Update start date before creation date test cases

sultanofcardio committed:
Use start time instead of end time for mobile column comparison logic

sultanofcardio committed:
Revert "Use start time instead of end time for mobile column comparison logic"

This reverts commit f08e73cfbb3025ff385bc9a8ed1da7bfbe0bb4e2.

sultanofcardio committed:
Change column name from item_hidden -> activity_or_flow_hidden

sultanofcardio committed:
Rename file from flow_item_history.csv -> flow_activity_history.csv

sultanofcardio committed:
Rename type property to activity_or_flow_hidden

sultanofcardio committed:
Rename flow_id and flow_name columns

sultanofcardio committed:
Add tests for findLatestMobileSchedule

sultanofcardio committed:
Merge branch 'refs/heads/develop' into feat/M2-8485-ema-data-exports

# Conflicts:
#	src/shared/types/featureFlags.ts

sultanofcardio committed:
Replace getWorkspaceRespondentsApi with apiDashboardSlice

sultanofcardio committed:
Implement filtering by activity and flow IDs

sultanofcardio committed:
Exclude history for users before they accept the applet invitation

sultanofcardio committed:
Fix compile issues

sultanofcardio committed:
Remove unused import

sultanofcardio committed:
Add missing default value for feature flag

sultanofcardio committed:
Use missing subject ID filters

sultanofcardio committed:
Stop downloading empty files again

The abstraction of the generateExportData method caused empty files to be downloaded again
